File: June 2025/20.py
from collections import Counter
import heapq

class Solution:
    def validgroup(self, arr, k):
        if len(arr) % k != 0:
            return False

        freq = Counter(arr)
        min_heap = list(freq.keys())
        heapq.heapify(min_heap)

        while min_heap:
            first = min_heap[0]
            for i in range(first, first + k):
                if freq[i] == 0:
                    return False
                freq[i] -= 1
                if freq[i] == 0 and i != min_heap[0]:
                    # Remove i from heap when it's no longer needed
                    min_heap.remove(i)
                    heapq.heapify(min_heap)
            # Remove the current min if its count reaches 0
            if freq[first] == 0:
                heapq.heappop(min_heap)

        return True

File: June 2025/20.cpp
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

class Solution {
  public:
    bool validgroup(vector<int> &arr, int k) {
        if (arr.size() % k != 0) return false;

        unordered_map<int, int> freq;
        for (int num : arr)
            freq[num]++;

        priority_queue<int, vector<int>, greater<int>> minHeap;
        for (auto &p : freq)
            minHeap.push(p.first);

        while (!minHeap.empty()) {
            int first = minHeap.top();

            for (int i = first; i < first + k; i++) {
                if (freq[i] == 0) return false;
                freq[i]--;
                if (freq[i] == 0 && i != minHeap.top()) {
                    // lazy removal: will get cleaned up on next round
                }
            }

            // Remove top if its count reached 0
            while (!minHeap.empty() && freq[minHeap.top()] == 0)
                minHeap.pop();
        }

        return true;
    }
};
